src/tools/base64.cpp

```diff
@@ -1,172 +1,119 @@
-
-  while (nprbytes > 4) {
-    *(bufout++) = (unsigned char)(pr2six[*bufin] << 2 | pr2six[bufin[1]] >> 4);
-    *(bufout++) =
-        (unsigned char)(pr2six[bufin[1]] << 4 | pr2six[bufin[2]] >> 2);
-    *(bufout++) = (unsigned char)(pr2six[bufin[2]] << 6 | pr2six[bufin[3]]);
-    bufin += 4;
-    nprbytes -= 4;
-  }
-
-  /* Note: (nprbytes == 1) would be an error, so just ingore that case */
-  if (nprbytes > 1) {
-    *(bufout++) = (unsigned char)(pr2six[*bufin] << 2 | pr2six[bufin[1]] >> 4);
-  }
-  if (nprbytes > 2) {
-    *(bufout++) =
-        (unsigned char)(pr2six[bufin[1]] << 4 | pr2six[bufin[2]] >> 2);
-  }
-  if (nprbytes > 3) {
-    *(bufout++) = (unsigned char)(pr2six[bufin[2]] << 6 | pr2six[bufin[3]]);
-  }
-
-  *(bufout++) = '\0';
-  nbytesdecoded -= (4 - nprbytes) & 3;
-  return nbytesdecoded;
-}
+    return opString;
+};
 
-static const char basis_64[] =
-    "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
-
-int Base64encode_len(int len) { return ((len + 2) / 3 * 4) + 1; }
-
-int Base64encode(char* encoded, const char* string, int len) {
-  int i;
-  char* p;
-
-  p = encoded;
-  for (i = 0; i < len - 2; i += 3) {
-    *p++ = basis_64[(string[i] >> 2) & 0x3F];
-    *p++ =
-        basis_64[((string[i] & 0x3) << 4) | ((int)(string[i + 1] & 0xF0) >> 4)];
-    *p++ = basis_64[((string[i + 1] & 0xF) << 2) |
-                    ((int)(string[i + 2] & 0xC0) >> 6)];
-    *p++ = basis_64[string[i + 2] & 0x3F];
-  }
-  if (i < len) {
-    *p++ = basis_64[(string[i] >> 2) & 0x3F];
-    if (i == (len - 1)) {
-      *p++ = basis_64[((string[i] & 0x3) << 4)];
-      *p++ = '=';
-    } else {
-      *p++ = basis_64[((string[i] & 0x3) << 4) |
-                      ((int)(string[i + 1] & 0xF0) >> 4)];
-      *p++ = basis_64[((string[i + 1] & 0xF) << 2)];
+std::string Base64::base64decode(uint8_t const* encodedString, 
+                                 const size_t len)
+{
+    if ((encodedString == NULL) || (len == 0) ) {
+        spdlog::error("Invalid data");
+        return std::string();
+    }
+    
+    size_t pos = 0;
+    std::string result;
+    result.reserve(len / 4 * 3);
+    constexpr int decodeWordLen = 4;
+
+    while (pos < len) {
+        size_t posOfChar1 = posOfChar(encodedString[pos+1]);
+        result.push_back(static_cast<std::string::value_type>( ( (posOfChar(encodedString[pos+0]) << 2 ) + ( (posOfChar1 & 0x30 ) >> 4))));
+
+        if ( ( pos + 2 < len ) &&  encodedString[pos + 2] != '=' ) {
+            uint32_t posOfChar2 = posOfChar(encodedString[pos + 2]);
+            result.push_back(static_cast<std::string::value_type>( (( posOfChar1 & 0x0f) << 4) + (( posOfChar2 & 0x3c) >> 2)));
+
+            if ( ( pos + 3 < len ) &&  (encodedString[pos + 3]) != '=' ) {
+                result.push_back(static_cast<std::string::value_type>( ( (posOfChar2 & 0x03 ) << 6 ) + posOfChar(encodedString[pos + 3]) ));
+            }
+        }
+
+        pos += decodeWordLen;
```

Comment:
When invalid array length is given (length> actual length), out of bound memory access would happen.
Add a try catch block to handle the exception.